### DISCUSSION TITLE
How to differentiate between JS "For....in" and "For...of"?

### DISCUSSION THREAD

**Discussion by esb** [ruby2js/ruby2js#208] [Q&A] [answered]
From my experiments, I seem to be able to only generate a JS "For...in" from a Ruby "For...in". Is there a way to get a "For...of" generated?

**Comment by jaredcwhite** [comment, marked as answer]
@esb The `functions` filter, which is in the default preset, automatically converts Ruby's `each` to `for...of`:

```ruby
arr = [1,2,3,4]

arr.each do |item|
  puts item
end
```

turns into

```js
let arr = [1, 2, 3, 4];

for (let item of arr) {
  console.log(item)
}
```

**Reply by esb** [reply]
Thanks for that. I've only just started usingf ruby2js, but I'm very impressed so far.

I'm trying to tame a massive JS system, and it's a nightmare, so hopefully ruby2js will allow me to bring some order to the chaos. JS is such a poorly designed language, and that makes it insanely hard to write clean code.
